content/posts/2022-09-14-introducing-jaiminho/index.md

```diff
@@ -0,0 +1,139 @@
+
+There are well-known methods to deal with such issues, and we chose to apply the transactional outbox pattern. Since we couldn’t find any related solution on [Django community](https://djangopackages.org/search/?q=outbox), we decided to implement an open-source library called [Jaiminho](https://github.com/loadsmart/jaiminho), which is a broker agnostic implementation of the outbox pattern.
+
+
+# The Transactional outbox pattern
+
+For simplicity purposes, let’s assume the common case of an application that has to save data to a database and notify it to an event stream. The idea behind the transactional outbox pattern is combining both writes into a single atomic transaction, which guarantees that if one of them fails, the entire transaction fails, avoiding inconsistent scenarios; in this case, it means that if a record is persisted into the database, it is guaranteed to be notified to the stream. The pattern achieves this goal by introducing two concepts:
+
+- Outbox table
+
+When storing a record to the database, the corresponding event should be persisted into an outbox table, and both operations should be in the same transaction. With this strategy we can leverage the atomicity property of relational databases transactions to ensure either both or no inserts will be successful.
+
+- Message relay
+
+The outbox table is guaranteed to have all events that should be notified, so we just have to ensure that they are really sent. This responsibility goes to the message relay, which will publish all events in the table to the stream and mark them as sent.
+Note: Since the message relay has to send an event and then mark it as sent, that characterizes another dual write, because if the relay fails to update the status after sending the event, it will be sent again on the next execution. So, to ensure no duplicated events are consumed, the message consumers must implement idempotence on their side.
+
+# Using Jaiminho in your project
+
+To add Jaiminho to your project, you have to:
+
+1) Install it:
+
+```console
+python -m pip install jaiminho
+```
+
+2) Add it to the Django project’s `INSTALLED_APPS`:
+
+```python
+INSTALLED_APPS = [
+    "django.contrib.auth",
+    "django.contrib.contenttypes",
+    ...
+    ...
+    ...
+    "jaiminho"
+]
+```
+
+3) Configure the options in the project’s Django `settings.py`:
+
+```python
+JAIMINHO_CONFIG = {
+    "PERSIST_ALL_EVENTS": False,
+    "DELETE_AFTER_SEND": True,
+    "PUBLISH_STRATEGY: "keep-order"
+    }
+```
+
+4) Add the `@save_to_outbox` decorator to any method responsible for communicating with external systems (brokers, external APIs, etc):
+
+```python
+from jaiminho.send import save_to_outbox
+
+@save_to_outbox
+def any_external_call(**kwargs):
+    # do something
+    return
+```
+
```

Comment:
**suggestion**: If you move this section to the end, I think you could add one more item ( `5)`) giving an example of how to configure a message relay.

By this way, seems that Jaiminho will do it automatically.

Reply:
Nice! I am thinking about adding an example using the run_in_loop flag, what do you think?

Reply:
Great idea!

Reply:
Done!